feat: implement `read`

Rich-Harris opened the pull request:
This implements the `read(file: string): Response` idea proposed in https://github.com/sveltejs/kit/pull/11647#issuecomment-1894442101.

Adapters can provide an implementation when they call `server.init`:

```js
server.init({
  env,
  read: (file) => new ReadableStream(...)
});
```

This also adds a `read` function (name bikesheddable) in `@sveltejs/kit/node` that turns a filepath into a `Response`, meaning that Node-like adapters have an easy job (see the `adapter-node` example in this PR).

Files that fall below the `assetsInlineLimit` are also handled (adapters don't need to provide an implementation, it's handled inside Kit).

TODO:

- [x] implement for other adapters (namely `adapter-netlify` and `adapter-vercel`)
- [x] docs
- [x] bikeshed the name
- [x] figure out how this all interacts with `paths.base`
- [x] add server assets imported by `hooks.server.js`
- [x] ensure `$app/server` can only be imported by server code
- [x] throw during dev/build if `read` from `$app/server` is used, but adapter provides no `read` implementation

---

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### Edits

- [x] Please ensure that 'Allow edits from maintainers' is checked. PRs without this option may be closed.

Conduitry commented:
Should it be the responsibility of each adapter's `readAsset` function to check that it's not being asked to read some potentially sensitive file that lives outside where the assets are supposed to live? Or is the idea that if you call `readAsset` on anything other than something generated at build time by Vite, then you deserve whatever you get?

Conduitry commented on packages/kit/src/exports/node/index.js:
```diff
@@ -189,3 +192,22 @@ export async function setResponse(res, response) {
+ * @param {string} file
+ */
+export function readFile(file) {
+	const stats = fs.statSync(file);
```
The `statSync` makes me uncomfortable, but I realize that the alternative to that is for each adapter's `readAsset` to be async.

Do we want to do any prod-only caching of asset stats? I'm not entirely certain what sorts of use cases we're trying to enable here, exactly. Do we expect that someone may call `readAsset` many times for the same path? Do we hope that they don't do that?

Rich-Harris replied:
What is it about `statSync` that makes you nervous?

Conduitry replied:
It being synchronous and thus blocking any other concurrent requests to the server while we wait for the filesystem.

Rich-Harris commented:
> Should it be the responsibility of each adapter's `readAsset` function to check that it's not being asked to read some potentially sensitive file that lives outside where the assets are supposed to live?

I think that probably falls under 'play stupid games, win stupid prizes'. I suppose we could check the input against the asset list (which gets shipped with the app anyway IIRC).

Another thought: if we _did_ do that, we could make `statSync` unnecessary by passing the content length (and content type) to the `readAsset` implementation...

Conduitry commented:
Oh, I like the idea of `readAsset` being passed the metadata which is already in-memory, rather than requiring it to hit the disk.

Rich-Harris commented:
Implemented that idea. I also changed the API as a result — if Kit already knows the length and type, there's no point passing that information to the adapter-provided `readAsset` implementation just so it can be returned in the form of response headers. Instead, `readAsset` has a simpler job, namely converting a filepath into a `ReadableStream`.

As a bonus, we're now able to throw an error if `readAsset` is called with something other than a Vite-generated asset path.

There's some room for improvement in the implementation but I feel good about this direction.

Rich-Harris commented:
Alright, got it working on Vercel as well as Node (and possibly Netlify, though I don't know how to test it without cutting a release — the manual deploy feature doesn't seem to know what to do with a `.netlify/functions-internal` directory).

Definitely room for some bikeshedding, but that can wait until the morning.

I haven't tested it, but there's no reason this shouldn't work with `import.meta.glob` — that'll be helpful for things like docs, in cases where prerendering isn't suitable for whatever reason.

dummdidumm commented on packages/adapter-vercel/index.js:
```diff
@@ -77,12 +79,9 @@ const plugin = function (defaults = {}) {
-				);
+				await create_function_bundle(builder, `${tmp}/index.js`, dir, config);
+
+				builder.writeServerAssets(dir, routes);
```
```suggestion
				builder.writeServerAssets?.(dir, routes); // TODO use optional chaining once peer dep contains a minimum SvelteKit version that has the method
```

dummdidumm reviewed (commented):
Makes sense overall 👍

Rich-Harris commented on packages/kit/src/exports/public.d.ts:
```diff
@@ -1156,6 +1164,7 @@ export interface SSRManifest {
 	/** private fields */
 	_: {
 		client: NonNullable<BuildData['client']>;
+		files: Record<string, [length: number, type: string]>;
```
This is wrong I think

Rich-Harris commented on packages/kit/src/runtime/app/server/index.js:
```diff
@@ -0,0 +1,43 @@
+ * @param {string} file
+ * @returns {Response}
+ */
+export function readAsset(file) {
```
I'm not sure about this name. `assets` is used to mean the contents of `static` as well as the build output. Not sure what would be better

Rich-Harris replied:
maybe just `read`

dummdidumm commented on packages/adapter-netlify/src/serverless.js:
```diff
@@ -1,6 +1,7 @@
 import './shims';
 import { Server } from '0SERVER';
 import { split_headers } from './headers.js';
+import { createReadable } from '@sveltejs/kit/node';
```
Either the adapters all need major version bumps or you do `import * as node_helpers` so that you can access it defensively below

benmccann commented on packages/kit/types/index.d.ts:
```diff
@@ -1873,6 +1884,10 @@ declare module '@sveltejs/kit/node' {
 	export function setResponse(res: import('http').ServerResponse, response: Response): Promise<void>;
+	/**
+	 * Converts a file on disk to a readable stream
+	 * */
```
```suggestion
	 */
```

benmccann commented on packages/kit/types/index.d.ts:
````diff
@@ -2107,6 +2122,21 @@ declare module '$app/paths' {
+	 * const asset = read(somefile);
+	 * const text = await asset.text();
+	 * ```
+	 * */
````
```suggestion
	 */
```

benmccann commented on packages/kit/types/index.d.ts:
```diff
@@ -78,6 +78,11 @@ declare module '@sveltejs/kit' {
 
+		/**
+		 * Find all the assets imported by server files belonging to `routes`
+		 * */
```
```suggestion
		 */
```

benmccann commented:
Assuming this would work with any file, I think I probably prefer `readFile` as a name since "asset" already has a particular connotation in Vite/Rollup

Rich-Harris commented:
It might be good if we can detect early on whether `read` is supported for a given route. Perhaps we need something like this:

```js
export default function adapter(opts) {
  return {
    name: '@sveltejs/adapter-vercel',
    async adapt(builder) {...},
    supports: {
      read: (config) => config.runtime !== 'edge'
    }
  };
}
```

This would allow us to error early, both at dev time and when building, if a given dynamic route would fail in production if it uses `read`. At dev time we'd presumably need to use `AsyncLocalStorage` to associate a given `read` call with a specific route.

Rich-Harris commented:
Alright, we now have adapter-based feature checking at dev and build time. It's a bit tricksy so I'll briefly explain how it works:

Inside the `read` function exported from `$app/server`, there is this line:

```js
__SVELTEKIT_TRACK__('$app/server:read');
```

During development, `__SVELTEKIT_TRACK__` is a global. When called, it grabs the current `RequestEvent` and route `config` from async local storage, and calls the relevant `supports.foo` function on the adapter. So far, there's just one tracked feature — `$app/server:read` — which we detect support for with `adapter.supports.read(...)`.

At build time, Rollup leaves the `__SVELTEKIT_TRACK__` call untouched, because it looks like a global. This allows us to figure out which output chunk, if any, contains the function in question (if it's not used anywhere, the function will be treeshaken away, and no chunk will contain it). Once found, we replace it with a comment.

Later, we can use SvelteKit's manifest data, combined with the Vite server manifest, to figure out which routes import that chunk. (There _is_ a danger of false positives here, I just realised — if route A imports `read` from `$app/server` and route B imports something else from `$app/server`, route B will be flagged as using `read` even though it doesn't. We could fix that with more sophisticated static analysis in future, if that ends up happening.)

We then use the same feature detection logic as in dev.

This is admittedly quite complex! But I'm not sure that there's a better alternative — it would suck if it just failed in production.

benmccann commented on packages/kit/src/exports/node/index.js:
```diff
@@ -189,3 +191,12 @@ export async function setResponse(res, response) {
+ * @param {string} file
+ * @returns {ReadableStream}
+ */
+export function createReadable(file) {
```
can we call this something else? I wasn't sure what it was from the name. Maybe `toStream` or `toReadableStream`?

Rich-Harris replied:
`toFoo` sort of implies that it's converting from something to something else (like the `Readable.toWeb` function that it uses under the hood), so I prefer `create`. It can be `createReadableStream` or something if that's preferable though

benmccann replied:
that's fair. `createReadableStream` would be fine with me. it's a little long, but I think that's fine since it's not used super frequently and adds some clarity

Rich-Harris commented:
There's one small thing left to do around detecting use of `read` inside `hooks.server.js`, but it's a small thing and I'm signing off for the night, so I'm going to mark this ready for review

Oh, it needs changesets as well I guess. Ready for review apart from that

benmccann commented on packages/adapter-netlify/index.js:
```diff
@@ -92,6 +93,19 @@ export default function ({ split = false, edge = edge_set_in_env_var } = {}) {
+			read: ({ route }) => {
+				if (edge) {
+					throw new Error(
+						`${name}: Cannot use \`read\` from \`$app/server\` in route \`${route.id}\` when using edge functions`
```
should we bother mentioning the route ID here or is that just misleading? whether it's edge or not doesn't depend on the route currently

Rich-Harris replied:
that was my thinking, yeah — it doesn't matter what the route is

benmccann commented on .changeset/rude-apples-jam.md:
```diff
@@ -0,0 +1,7 @@
+'@sveltejs/adapter-node': major
+---
+
+breaking: update peer dependency on `@sveltejs/kit`
```
```suggestion
breaking: update peer dependency on `@sveltejs/kit` and add support for reading assets from filesystem
```

benmccann commented on packages/adapter-vercel/index.js:
```diff
@@ -77,12 +80,15 @@ const plugin = function (defaults = {}) {
-				);
+				await create_function_bundle(builder, `${tmp}/index.js`, dir, config);
+
+				// TODO remove `if` once peer dep contains a minimum SvelteKit version that has the method
```
I think this could be addressed now since we're bumping the peer dep

benmccann commented on packages/kit/src/core/generate_manifest/find_server_assets.js:
```diff
@@ -0,0 +1,53 @@
+import { find_deps } from '../../exports/vite/build/utils.js';
+
+/**
+ * Generates the data used to write the server-side manifest.js file. This data is used in the Vite
```
"generates" makes me think it's doing more than creating an in-memory representation
```suggestion
 * Returns the data used to write the server-side manifest.js file. This data is used in the Vite
```

Rich-Harris replied:
I think I might have just copy-pasted this from elsewhere, the description is totally wrong

benmccann commented on packages/kit/src/core/adapt/builder.js:
```diff
@@ -144,6 +145,13 @@ export function create_builder({
 			}
 		},
 
+		findServerAssets(route_data) {
```
should this have a jsdoc since it's public? there's one added on `find_server_assets`, but people might not be able to easily see that one

Rich-Harris replied:
it has one, the `Builder` interface is described in `public.d.ts` https://github.com/sveltejs/kit/pull/11649/files#diff-a174e22e6d675073a963705b97687d42e219c2e05eb1cd6d5811c2581d416a8eR110-R113

benmccann commented on packages/kit/src/exports/public.d.ts:
```diff
@@ -89,13 +99,20 @@ export interface Builder {
 
+	/**
+	 * Find all the assets imported by server files belonging to `routes`
+	 * @param routes
```
I don't think this line does much since there's no description and the types are already present below
```suggestion
```

benmccann commented on packages/kit/types/index.d.ts:
```diff
@@ -1873,6 +1900,10 @@ declare module '@sveltejs/kit/node' {
+	/**
+	 * Converts a file on disk to a readable stream
+	 * */
+	export function createReadable(file: string): ReadableStream;
```
```suggestion
	export function createReadableStream(file: string): ReadableStream;
```

benmccann reviewed (commented):
Pieces of your comment above (https://github.com/sveltejs/kit/pull/11649#issuecomment-1896929650) should probably make its way into some code comments probably split across some of the files in `src/exports/vite`

benmccann commented on packages/kit/src/exports/public.d.ts:
```diff
@@ -35,6 +35,16 @@ export interface Adapter {
+	/**
+	 * Functions called during dev and build to determine whether specific features will work in production with this adapter
+	 */
+	supports?: {
```
should we add something about this to https://kit.svelte.dev/docs/writing-adapters ?  our docs there are pretty limited, but could keep us from falling further behind

Rich-Harris replied:
done

benmccann commented on packages/kit/src/exports/public.d.ts:
```diff
@@ -35,6 +35,16 @@ export interface Adapter {
 	 */
 	adapt(builder: Builder): MaybePromise<void>;
+	/**
+	 * Functions called during dev and build to determine whether specific features will work in production with this adapter
```
I don't know if it's premature, but I imagine not everything would be a function in the future and you could also just have some simple booleans for some things. E.g. You might have `ssr: false` for `adapter-static`
```suggestion
	 * Checks called during dev and build to determine whether specific features will work in production with this adapter
```

benmccann reviewed (commented):
we should probably mention in the docs besides just the API docs. I'm not sure where though. https://kit.svelte.dev/docs/server-only-modules could potentially be an option if tweaked a bit since this is something that works only on the server

benmccann commented on sites/kit.svelte.dev/src/routes/content.json/content.server.js:
```diff
@@ -1,13 +1,11 @@
 } from '@sveltejs/site-kit/markdown';
-import { readFile } from 'node:fs/promises';
-import path from 'node:path';
-import glob from 'tiny-glob';
```
I think you can remove `tiny-glob` from the `package.json` now

dominikg commented on packages/kit/src/exports/public.d.ts:
```diff
@@ -1159,6 +1178,8 @@ export interface SSRManifest {
 		routes: SSRRoute[];
 		matchers(): Promise<Record<string, ParamMatcher>>;
+		/** A `[file]: size` map of all assets imported by server code */
+		server_assets: Record<string, number>;
```
if someone used a glob import, would this potentially turn into a very large list? do we actually need all filenames if only few might ever be passed through `read`?

Rich-Harris replied:
Yes but this is just a caveat of using `import.meta.glob` _anyway_ — you will always end up with filenames written into the resulting bundle, so you'd better make sure you're only globbing things that you will actually use

Rich-Harris commented on packages/kit/src/core/postbuild/analyse.js:
```diff
@@ -163,30 +185,59 @@ function analyse_page(layouts, leaf) {
+	}
+
+	if (manifest_data.hooks.server) {
+		// TODO if hooks.server.js imports `read`, it will be in the entry chunk
```
Due to some quirk of how the resulting Vite bundle is structured (or something like that) I couldn't actually get this to work reliably, so I kicked the can down the road for now. The reason I felt comfortable doing that is that if you used `read` inside `handle` you would get a dev time error on every request (and if it was in `hooks.server.js` _outside_ handle then the app wouldn't start in the first place), so the chance of this causing an unexpected error in prod is effectively zero

dominikg commented on packages/kit/src/exports/vite/dev/index.js:
```diff
@@ -27,6 +29,15 @@ const cwd = process.cwd();
 export async function dev(vite, vite_config, svelte_config) {
 	installPolyfills();
 
+	const async_local_storage = new AsyncLocalStorage();
```
this has been added in node 18.16 (if the node homepage is to believed) but kit currently has "node": ">=18.13" in its package.json. So technically a breaking change

Rich-Harris replied:
Where are you seeing that? This is from https://nodejs.org/api/async_context.html#class-asynclocalstorage

<img width="624" alt="image" src="https://github.com/sveltejs/kit/assets/1162160/a4514ba0-a7bd-4908-834c-9a449f4b1b1c">

dominikg replied:
didn't expand the history and mistook the information for bind below for the class...

benmccann commented on packages/kit/src/exports/vite/index.js:
```diff
@@ -685,6 +705,18 @@ async function kit({ svelte_config }) {
+		renderChunk(code, chunk) {
+			if (code.includes('__SVELTEKIT_TRACK__')) {
+				return {
+					code: code.replace(/__SVELTEKIT_TRACK__\('(.+?)'\)/g, (_, label) => {
```
it could help to add a comment saying what the code is that's being output and what it does

Rich-Harris replied:
feels kinda duplicative with the code

benmccann commented on pnpm-lock.yaml:
```diff
@@ -289,7 +289,7 @@ importers:
       '@sveltejs/kit':
         specifier: ^1.0.0 || ^2.0.0
-        version: link:../kit
+        version: 2.3.5(@sveltejs/vite-plugin-svelte@3.0.1)(svelte@4.2.8)(vite@5.0.11)
```
hmm. this is weird

benmccann commented on packages/kit/src/types/internal.d.ts:
```diff
@@ -120,12 +126,18 @@ export class InternalServer extends Server {
 		options: RequestOptions & {
 			prerendering?: PrerenderOptions;
 			read: (file: string) => Buffer;
+			before_handle?: (event: RequestEvent, config: any, prerender: PrerenderOption) => void;
```
this could use a comment explaining it's a hook called before handle

Rich-Harris replied:
```suggestion
			/** A hook called before `handle` during dev, so that `AsyncLocalStorage` can be populated */
			before_handle?: (event: RequestEvent, config: any, prerender: PrerenderOption) => void;
```

Rich-Harris commented on documentation/docs/25-build-and-deploy/99-writing-adapters.md:
```diff
@@ -19,6 +20,13 @@ export default function (options) {
+		supports: {
+			read: ({ config, route }) => {
+				// return `true` if the route with the given `config` can use `read`
+				// from `$app/server` in production, return false if it can't.
```
it's meant to be two sentences

```suggestion
				// from `$app/server` in production, return `false` if it can't.
```

Rich-Harris commented on packages/kit/src/runtime/app/server/index.js:
```diff
@@ -0,0 +1,58 @@
+	const file = DEV && asset.startsWith('/@fs') ? asset : asset.slice(base.length + 1);
+
+	if (file in manifest._.server_assets) {
+		const length = manifest._.server_assets[file];
```
no. if we did that, then the `read` implementation passed to `server.init` would need to return the length as well as the body `ReadableStream` — it makes the API needlessly convoluted, just so we can avoid statting the file (a very cheap operation) at build time (when it doesn't matter)

Rich-Harris commented on packages/kit/src/exports/vite/index.js:
```diff
@@ -224,6 +225,9 @@ async function kit({ svelte_config }) {
 	const service_worker_entry_file = resolve_entry(kit.files.serviceWorker);
 	const parsed_service_worker = path.parse(kit.files.serviceWorker);
 
+	/** @type {Record<string, string[]>} */
```
```suggestion
	/**
	 * A map showing which features (such as `$app/server:read`) are defined
	 * in which chunks, so that we can later determine which routes use which features
	 * @type {Record<string, string[]>}
	 */
```

benmccann commented on packages/kit/src/exports/vite/index.js:
```diff
@@ -685,6 +709,18 @@ async function kit({ svelte_config }) {
+				return {
+					code: code.replace(/__SVELTEKIT_TRACK__\('(.+?)'\)/g, (_, label) => {
+						(tracked_features[chunk.name + '.js'] ??= []).push(label);
+						return `/* track ${label}            */`;
```
assuming I've surmised correctly why the spaces are there...
```suggestion
						// put extra whitespace at the end of the comment to preserve the source size and avoid interfering with source maps
						return `/* track ${label}            */`;
```

gitaarik commented:
This implementation has some limitations:

- Hard to generate HTTP ETAG header for smart caching
- No support for data: images that are not base64 encoded, like SVGs with just raw XML

I made a this gist as a work-around, which can be used for inspiration of a more flexible implementation in SvelteKit:

https://gist.github.com/gitaarik/59213791bc2e10977de81d90ac5a0fdb
